src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSession.cs

```diff
@@ -28,7 +34,11 @@ public ProjectHotReloadSession(
         Lazy<IHotReloadAgentManagerClient> hotReloadAgentManagerClient,
         Lazy<IHotReloadDiagnosticOutputService> hotReloadOutputService,
         Lazy<IManagedDeltaApplierCreator> deltaApplierCreator,
-        IProjectHotReloadSessionCallback callback)
+        IProjectHotReloadSessionCallback callback,
+        IProjectHotReloadSessionManager? sessionManager = null,
+        IDebugLaunchProvider? launchProvider = null,
+        ILaunchProfile? launchProfile = null,
+        DebugLaunchOptions? debugLaunchOptions = null)
```

Comment:
Are these only populated on the project that was launched? I'm assuming we create an instance of `ProjectHotReloadSession` for all projects.

Is the expectation that these three values would either all be provided, or all be null? If so, perhaps wrap them in a new type with non-null fields, so that the null check only happens once. It might be clearer.

Reply:
> Are these only populated on the project that was launched? I'm assuming we create an instance of ProjectHotReloadSession for all projects.

It's created for project that get launched.

> Is the expectation that these three values would either all be provided, or all be null? If so, perhaps wrap them in a new type with non-null fields, so that the null check only happens once. It might be clearer.

Sure